src/WebJobs.Script.WebHost/Metrics/FlexConsumptionMetricsPublisher.cs

```diff
@@ -0,0 +1,301 @@
+        /// </summary>
+        internal void Initialize()
+        {
+            _metricPublishInterval = TimeSpan.FromMilliseconds(_optionsMonitor.CurrentValue.MetricsPublishIntervalMS);
+            _initialPublishDelay = TimeSpan.FromMilliseconds(_optionsMonitor.CurrentValue.InitialPublishDelayMS);
+            _intervalStopwatch = ValueStopwatch.StartNew();
+            MetricsFilePath = _optionsMonitor.CurrentValue.MetricsFilePath;
+
+            IsAlwaysReady = _environment.GetEnvironmentVariable(EnvironmentSettingNames.FunctionsAlwaysReadyInstance) == "1";
+        }
+
+        internal async Task OnPublishMetrics()
+        {
+            try
+            {
+                if (FunctionExecutionCount == 0 && FunctionExecutionTimeMS == 0 && !IsAlwaysReady)
+                {
+                    // no activity to report
+                    return;
+                }
+
+                // we've been accumulating function activity for the entire period
+                // publish this activity and reset
+                Metrics metrics = null;
+                lock (_lock)
+                {
+                    metrics = new Metrics
+                    {
+                        TotalTimeMS = (long)_intervalStopwatch.GetElapsedTime().TotalMilliseconds,
+                        ExecutionCount = FunctionExecutionCount,
+                        ExecutionTimeMS = FunctionExecutionTimeMS,
+                        IsAlwaysReady = IsAlwaysReady
+                    };
+
+                    FunctionExecutionTimeMS = FunctionExecutionCount = 0;
+                }
+
+                await PublishMetricsAsync(metrics);
+            }
+            finally
+            {
+                _intervalStopwatch = ValueStopwatch.StartNew();
+            }
+        }
+
+        private async void OnFunctionMetricsPublishTimer(object state)
+        {
+            await OnPublishMetrics();
+        }
+
+        private async Task PublishMetricsAsync(Metrics metrics)
+        {
+            string fileName = string.Empty;
+
+            try
+            {
+                bool metricsPublishEnabled = !string.IsNullOrEmpty(MetricsFilePath);
+                if (metricsPublishEnabled && !PrepareDirectoryForFile())
+                {
+                    return;
+                }
+
+                string metricsContent = JsonConvert.SerializeObject(metrics);
+                _logger.PublishingMetrics(metricsContent);
+
+                if (metricsPublishEnabled)
+                {
+                    fileName = $"{Guid.NewGuid().ToString().ToLower()}.json";
+                    string filePath = Path.Combine(MetricsFilePath, fileName);
+
+                    using (var streamWriter = _fileSystem.File.CreateText(filePath))
+                    {
+                        await streamWriter.WriteAsync(metricsContent);
+                    }
+                }
```

Comment:
>} [](http://example.com/codeflow?start=16&length=1)

since we know this publisher is only enabled if required and if metricsPublishEnabled is not true then it should be a clear signal of an error 

may be we can add an else block and an error log here.
I think we should also incorporate this failed state in the health check endpoint (but that discussion can be part of health checks).

Reply:
I chose to let the publisher to run in this mode to allow us to phase this into the runtime w/o breaking changes or starting to generate a bunch of errors for Flex Apps before the rest of the pipeline is hooked up.

Note that during startup above I do already log MetricsFilePath, so we can write queries to look for problems. We'll also have alerting for production apps.

If that's not sufficient, perhaps we add the error later after everything else is done?

Reply:
as long as we have something we can use to query/alert, we should be good.